In [1]:
import pandas as pd
import numpy as np
import jieba
import re
dataset=pd.read_csv("train.csv")
dataset.columns=['id','content','comment_all','label']
dataset=dataset.set_index('id')
dataset

content  \
id                                                         
9982   \t\t#王嘉尔##热血街舞团# 王嘉尔或因加盟爱奇艺《热血街舞团》遭优酷全面封杀，目前优酷...   
9983   谁的群最多，帮忙转一下，找一名弥度职中学生，叫刘冻雪，16岁，请速回鲁甸县医院，妈妈伤的很严...   
9984   \t\t妈蛋，以色列在战术层面的这些神操作真是令人叹为观止 以战机以俄罗斯战机为掩护并释放电...   
9985   帮朋友转的，急找孩子，求转，求帮忙寻人启事13940292999。有线索酬金10万，今天上午...   
9986   \t\t#不情愿的新娘# 看的吻戏最多的一部泰剧从第一集到第四集数不清哈哈哈哈虽然是沙雕剧情...   
...                                                  ...   
49905  \t\t网传是实习生误删…这个我是不信的（实习生哪来的权限），希望不要有太多技术人员背锅//...   
49906  北京时间3月12日消息，在83版《西游记》中扮演孙悟空的演员六小龄童（章金莱），3月12日早...   
49907                             学生举报学校雾霾天上课却被教育局泄露电话号码   
49908  \t\t惊！国民儿子某蛙光天化日之下脚踩两只船！对此，记者采访到近日刚刚结束电影《鼠来宝3》...   
49909  \t  &#xe627; 张继科 小周：偷偷告诉你们，继科对签名照的照片很满意拿走了几张（觉...   

                                             comment_all  label  
id                                                               
9982   不想理你，大家不要供献热度，谣言止于智者\t优酷和汪涵合作深入，而何炅一直特别挺王嘉尔，这…...      0  
9983                                                 NaN     -1  
9984                                                 NaN      0  
9985   你受骗上当了，这是虚假消息！请看辟谣网址（O谣场现形记）。尽快删除帖子，因为已被举报。\t回...     -1  
9986                                  后面要虐[泪]\t都在哪里看的?\t      0  
...                                                  ...    ...  
49905                                                NaN      0  
49906  这是假消息的嘛\t回复@顔鋭:真的\t[蜡烛][蜡烛]1[蜡烛]2\t是假的，当事人还在网上...     -1  
49907                                                NaN      1  
49908                                                NaN      0  
49909  小周是谁[允悲]1[允悲]\t哈哈哈，想不到吧[允悲]\t你够啦@张继科;;[挖鼻]1[挖鼻...      0  

[39928 rows x 3 columns]

以上是初始数据集

In [2]:
test=pd.read_csv("test.csv")
test.columns=['id','content','comment_all']
test=test.set_index('id')
testset=test
testset

content  \
id                                                        
0     \t\t昨天在家楼下网点改了全渠道手机号，去单位楼下用专机表扬了大堂经理，并称不需要回复。今...   
1     【婆婆为维护出轨儿子虐待儿媳致其成”骷髅“】杨女士发现老公出轨，向婆婆反映。婆婆辩称“男人都...   
2     各位，有家长反映:今天中午放学时，有学生遇到开小车的向小学生问路，并拿出100元感谢费，让小...   
3     帮忙发一条救命信息。急需rh阴性血，救命之恩必有重谢。电话13845672513155616...   
4                             “逃犯克星”4立新功!张学友金华演唱会又有逃犯被抓   
...                                                 ...   
9977  \t  #历史影像# 中英街边界石碑上，殖民香港的英国警官与解放军对峙，美国杂志 LIFE ...   
9978  【开学季我们进校园】新年已过，转眼又到了孩子们开学的时候，为守护辖区师生校园安全，经济开发区...   
9979  一名中学生，叫周榆棍，17岁，请速回平山县医院，妈妈伤的很严重，想见她最后一面，爸爸号码：1...   
9980  艺龙真是恶心人，我在艺龙定了好几次宾馆，每次都说给我按照约定返现，可是一次都没有兑现，找各种...   
9981  学生家长捡到一个高考准考证，谁认识通知下，别耽误了高考大事。白娅倩考点一中考场013座号11...   

                                            comment_all  
id                                                       
0                                                   NaN  
1                                                   NaN  
2                                                   NaN  
3      此为虚假消息！请勿转发虚假消息！\t回复@花自漂零的天空:你知道什么你就举报我，闲在难受的啊\t  
4                                                   NaN  
...                                                 ...  
9977                                                NaN  
9978                                                NaN  
9979                                                NaN  
9980  您好，感谢反馈，会尽快与您联系\t的时候是问清楚酒店工作人员之后入住的，现在又来怪我？艺龙这...  
9981  @白较瘦@Ctrl键@大鹏看天下转发下吧，别耽误孩子高考[心]\t假消息\t吸金电话不要打\...  

[9982 rows x 2 columns]

dataset.loc[9983,'comment_all']\\
type(dataset.loc[9983,'comment_all'])\\
\\float

In [3]:
def clear(text)->str:
    #print(re.findall('[\u4E00-\u9FFF]+|\d',text))
    if type(text) is float:return 
    str1=""
    str2=str1.join(text)
    y=lambda x:re.findall('[\u4E00-\u9FFF]+',x)
    li=y(str2)
    for i in range(len(li)):
        li[i]=list(jieba.cut(li[i]))
    li2=[]
    for subli in li:
        for item in subli:
            li2.append(item)
    return li2
text='正则365表达\t式ds匹配/--*测\t试文字'
a=clear(dataset.loc[9982,'content'])
b=clear(dataset.loc[9984,'comment_all'])
c=clear(testset.loc[9979,'content'])
print(a)
print(b)
print(c)



Building prefix dict from the default dictionary ...
Loading model from cache /var/folders/34/5ls3qb4j0rl0kbsw8j22d1_w0000gn/T/jieba.cache
Loading model cost 0.441 seconds.
Prefix dict has been built successfully.


['王嘉尔', '热血', '街', '舞团', '王嘉尔', '或', '因', '加盟', '爱奇艺', '热血', '街', '舞团', '遭优酷', '全面', '封杀', '目前', '优酷', '视频', '已', '将', '王嘉尔', '的', '相关', '视频', '全面', '下架', '网传', '王嘉尔', '之前', '已', '答应', '录制', '优酷', '这', '就是', '街舞', '合同', '也', '已', '确定', '这次', '究竟', '是', '双方', '争夺', '艺人', '惹出', '事端', '还是', '王嘉尔', '出尔反尔', '你们', '怎么', '看']
None
['一名', '中学生', '叫', '周榆棍', '岁', '请', '速回', '平山县', '医院', '妈妈', '伤', '的', '很', '严重', '想见', '她', '最后', '一面', '爸爸', '号码', '爱心', '接力']


In [4]:
text="分词测试文字"
a=text.split()
a

['分词测试文字']

In [5]:
for index,row in dataset.iterrows():
    dataset.at[index,'content']=clear(dataset.at[index,'content'])
    dataset.at[index,'comment_all']=clear(dataset.at[index,'comment_all'])
dataset

content  \
id                                                         
9982   [王嘉尔, 热血, 街, 舞团, 王嘉尔, 或, 因, 加盟, 爱奇艺, 热血, 街, 舞团...   
9983   [谁, 的, 群, 最, 多, 帮忙, 转, 一下, 找, 一名, 弥度, 职, 中学生, ...   
9984   [妈蛋, 以色列, 在, 战术, 层面, 的, 这些, 神, 操作, 真是, 令人, 叹为观...   
9985   [帮, 朋友, 转, 的, 急, 找, 孩子, 求转, 求, 帮忙, 寻人启事, 有, 线索...   
9986   [不, 情愿, 的, 新娘, 看, 的, 吻, 戏, 最, 多, 的, 一部, 泰剧, 从,...   
...                                                  ...   
49905  [网传, 是, 实习生, 误删, 这个, 我, 是, 不信, 的, 实习生, 哪来, 的, ...   
49906  [北京, 时间, 月, 日, 消息, 在, 版, 西游记, 中, 扮演, 孙悟空, 的, 演...   
49907     [学生, 举报, 学校, 雾, 霾, 天, 上课, 却, 被, 教育局, 泄露, 电话号码]   
49908  [惊, 国民, 儿子, 某蛙, 光天化日, 之下, 脚, 踩, 两只, 船, 对此, 记者,...   
49909  [张继科, 小, 周, 偷偷, 告诉, 你们, 继科, 对, 签名, 照, 的, 照片, 很...   

                                             comment_all  label  
id                                                               
9982   [不想, 理, 你, 大家, 不要, 供献, 热度, 谣言止于, 智者, 优酷, 和, 汪涵...      0  
9983                                                None     -1  
9984                                                None      0  
9985   [你, 受骗上当, 了, 这是, 虚假, 消息, 请, 看, 辟谣, 网址, 谣场, 现形记...     -1  
9986                         [后面, 要虐, 泪, 都, 在, 哪里, 看, 的]      0  
...                                                  ...    ...  
49905                                               None      0  
49906  [这是, 假消息, 的, 嘛, 回复, 顔, 鋭, 真的, 蜡烛, 蜡烛, 蜡烛, 是, 假...     -1  
49907                                               None      1  
49908                                               None      0  
49909  [小周, 是, 谁, 允悲, 允悲, 哈哈哈, 想不到, 吧, 允悲, 你够, 啦, 张继科...      0  

[39928 rows x 3 columns]

用正则表达式来匹配汉字，其中\t作为分隔符应该单独分开？后面想了想，把数字和字符都给去掉了，只留下了汉字。

In [6]:
dataset2=[]
for index,row in dataset.iterrows():
    str2=" "
    if dataset.loc[index,'comment_all']is None:
        dataset2.append(dataset.loc[index,'content'])
    else:
        dataset2.append(dataset.loc[index,'content']+dataset.loc[index,'comment_all'])

In [7]:
dataset2

[['王嘉尔',
  '热血',
  '街',
  '舞团',
  '王嘉尔',
  '或',
  '因',
  '加盟',
  '爱奇艺',
  '热血',
  '街',
  '舞团',
  '遭优酷',
  '全面',
  '封杀',
  '目前',
  '优酷',
  '视频',
  '已',
  '将',
  '王嘉尔',
  '的',
  '相关',
  '视频',
  '全面',
  '下架',
  '网传',
  '王嘉尔',
  '之前',
  '已',
  '答应',
  '录制',
  '优酷',
  '这',
  '就是',
  '街舞',
  '合同',
  '也',
  '已',
  '确定',
  '这次',
  '究竟',
  '是',
  '双方',
  '争夺',
  '艺人',
  '惹出',
  '事端',
  '还是',
  '王嘉尔',
  '出尔反尔',
  '你们',
  '怎么',
  '看',
  '不想',
  '理',
  '你',
  '大家',
  '不要',
  '供献',
  '热度',
  '谣言止于',
  '智者',
  '优酷',
  '和',
  '汪涵',
  '合作',
  '深入',
  '而',
  '何炅',
  '一直',
  '特别',
  '挺',
  '王嘉尔',
  '这',
  '不好意思',
  '我',
  '在',
  '瞎',
  '想',
  '笑死人',
  '这个',
  '营销',
  '号',
  '说',
  '的',
  '自己',
  '好像',
  '亲眼看见',
  '了',
  '一样',
  '真签',
  '了',
  '合同',
  '优酷',
  '会',
  '不官',
  '宣',
  '吗',
  '谁',
  '会',
  '放过',
  '这么',
  '好',
  '的',
  '机会',
  '敢情',
  '营销',
  '号',
  '写',
  '东西',
  '都',
  '是',
  '这样',
  '乱写',
  '经不起',
  '推敲',
  '的',
  '费解',
  '允悲',
  '我',
  '觉得',
  '腾讯',
  '芒果',
  '爱奇艺',
  '站',
  '的',
  '视频',


In [8]:
for index,row in testset.iterrows():
    testset.at[index,'content']=clear(testset.at[index,'content'])
    testset.at[index,'comment_all']=clear(testset.at[index,'comment_all'])
testset

content  \
id                                                        
0     [昨天, 在家, 楼下, 网点, 改, 了, 全, 渠道, 手机号, 去, 单位, 楼下, ...   
1     [婆婆, 为, 维护, 出轨, 儿子, 虐待, 儿媳, 致其成, 骷髅, 杨, 女士, 发现...   
2     [各位, 有, 家长, 反映, 今天, 中午, 放学, 时, 有, 学生, 遇到, 开, 小...   
3     [帮忙, 发, 一条, 救命, 信息, 急需, 阴性, 血, 救命之恩, 必有, 重谢, 电...   
4           [逃犯, 克星, 立新功, 张学友, 金华, 演唱会, 又, 有, 逃犯, 被, 抓]   
...                                                 ...   
9977  [历史, 影像, 中英街, 边界, 石碑, 上, 殖民, 香港, 的, 英国, 警官, 与,...   
9978  [开学, 季, 我们, 进校园, 新年, 已过, 转眼, 又, 到, 了, 孩子, 们, 开...   
9979  [一名, 中学生, 叫, 周榆棍, 岁, 请, 速回, 平山县, 医院, 妈妈, 伤, 的,...   
9980  [艺龙, 真是, 恶心, 人, 我, 在, 艺龙定, 了, 好, 几次, 宾馆, 每次, 都...   
9981  [学生家长, 捡, 到, 一个, 高考, 准考证, 谁, 认识, 通知, 下, 别, 耽误,...   

                                            comment_all  
id                                                       
0                                                  None  
1                                                  None  
2                                                  None  
3     [此, 为, 虚假, 消息, 请勿, 转发, 虚假, 消息, 回复, 花自, 漂零, 的, ...  
4                                                  None  
...                                                 ...  
9977                                               None  
9978                                               None  
9979                                               None  
9980  [您好, 感谢, 反馈, 会, 尽快, 与, 您, 联系, 的, 时候, 是, 问, 清楚,...  
9981  [白, 较瘦, 键, 大鹏, 看, 天下, 转发, 下, 吧, 别, 耽误, 孩子, 高考,...  

[9982 rows x 2 columns]

In [9]:
dataset.to_pickle('split_dataset')
testset.to_pickle('split_testset')

先用tf-idf方法做出一个base-line，然后再考虑用transformer。
\\干，还是先想想词嵌入吧，CBOW和N-grams都不错。


path_translated='D:\\pycode\\jupyter\\dataset_news\\ChineseWordVectors\\evaluation\\testsets\\CA_translated\\ca_translated.txt'
path_morhological='D:\\pycode\\jupyter\\dataset_news\\ChineseWordVectors\\evaluation\\testsets\\CA8\morhological.txt'
path_word='D:\\pycode\\jupyter\\dataset_news\\ChineseWordVectors\\evaluation\\testsets\\wordset.txt'
filenames=[path_translated,path_morhological,path_semantic]
path_sogou='D:\\pycode\\jupyter\\dataset_news\\News'
with open(path_word,"r",encoding="utf-8") as f:
    text=f.readlines()
for str1 in text:
    str1.strip("\n")
    str1.split(" ")
text=[str1.strip("\n").split(" ")for str1 in text]
text

这里做了一个小demo。可以用得

In [10]:
from gensim.models import FastText

sentences：可以是一个list，对于大语料集，建议使用BrownCorpus,Text8Corpus或LineSentence构建。

sg： 用于设置训练算法，默认为0，对应CBOW算法；sg=1则采用skip-gram算法。

size：是指特征向量的维度，默认为100。大的size需要更多的训练数据,但是效果会更好. 推荐值为几十到几百。

window：表示当前词与预测词在一个句子中的最大距离是多少

alpha: 是学习速率

seed：用于随机数发生器。与初始化词向量有关。

min_count: 可以对字典做截断. 词频少于min_count次数的单词会被丢弃掉, 默认值为5

max_vocab_size: 设置词向量构建期间的RAM限制。如果所有独立单词个数超过这个，则就消除掉其中最不频繁的一个。每一千万个单词需要大约1GB的RAM。设置成None则没有限制。

sample: 高频词汇的随机降采样的配置阈值，默认为1e-3，范围是(0,1e-5)

workers参数控制训练的并行数。

hs: 如果为1则会采用hierarchical softmax技巧。如果设置为0（defaut），则negative sampling会被使用。

negative: 如果>0,则会采用negativesamping，用于设置多少个noise words

cbow_mean: 如果为0，则采用上下文词向量的和，如果为1（defaut）则采用均值。只有使用CBOW的时候才起作用。

hashfxn： hash函数来初始化权重。默认使用python的hash函数

iter： 迭代次数，默认为5

trim_rule： 用于设置词汇表的整理规则，指定那些单词要留下，哪些要被删除。可以设置为None（min_count会被使用）或者一个接受()并返回RU·E_DISCARD,uti·s.RU·E_KEEP或者uti·s.RU·E_DEFAU·T的函数。

sorted_vocab： 如果为1（defaut），则在分配word index 的时候会先对单词基于频率降序排序。

batch_words：每一批的传递给线程的单词的数量，默认为10000

In [11]:
model2=FastText(dataset2, sg=1,vector_size=100, window=5, min_count=1, workers=4)

model2.wv.most_similar("吴官骏")

print(model2.wv.similarity('陈逸', '扶贫'))

len(model2.wv['代码'])

接下来要做的事：整理成语料库，然后放入txt文档，然后放入model训练。生成预训练模型。
<p>数据训练部分已经完成。看上面几行示例就可以获得相应的词向量。接下来完成Embedding部分。

In [12]:
import numpy as np
c=[]
for sequence in dataset2:
    b=[]
    for word in sequence:
        a=model2.wv[word]
        b.append(a)
    c.append(b)
data_numpy=np.array(c)


/var/folders/34/5ls3qb4j0rl0kbsw8j22d1_w0000gn/T/ipykernel_33188/738509728.py:9: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
  data_numpy=np.array(c)


In [13]:
data_numpy

array([list([array([ 0.01111047,  0.15573478,  0.5394658 ,  0.5776898 , -0.20704894,
       -0.53447825, -0.27937084,  0.6725365 , -0.0279234 ,  0.351369  ,
        0.10360638, -0.26450968,  0.05848536,  0.37791204,  0.40422088,
        0.23985182,  0.31472486, -0.6536147 , -0.47137547, -0.8082617 ,
       -0.1389393 ,  0.03160868,  0.15291962,  0.4206323 , -0.21481644,
        0.42859402, -0.39408174, -0.28464338, -0.11427695,  0.6978684 ,
        0.6890319 , -0.6084487 ,  0.04778454,  0.04752538, -0.6574274 ,
        0.24576247,  0.10789709, -0.20991416, -0.19647367, -0.5293797 ,
        0.06690597,  0.36516914, -0.42938083,  0.841566  , -0.14935076,
        0.54856694, -0.52987754,  0.09245888, -0.23080194, -0.41130668,
        0.5200139 , -0.21514294, -0.71834755,  0.12149673,  0.6412147 ,
        0.2992021 ,  0.1267975 , -0.19855747, -0.39989787, -0.00509983,
        0.22837617,  0.517133  , -0.48282552,  0.04276463, -0.2738971 ,
        0.76294523,  0.24133424,  0.15804137, -0.69